File: bindings/csharp/examples/ToolInvocation.cs
/// Tool invocation: register a tool with test vectors and invoke it.

using Limn.Scp;

await using var identity = await Identity.CreateAsync(custody: "platform");

await using var ctx = await Context.CreateAsync(
    identity,
    new ContextParams
    {
        Ceiling = ["msg:send", "msg:receive", "tool:invoke"],
        Tools =
        [
            new ToolDefinition(
                Name: "weather",
                Description: "Get current weather for a city",
                InputSchema: new Dictionary<string, object>
                {
                    ["type"] = "object",
                    ["properties"] = new Dictionary<string, object>
                    {
                        ["city"] = new Dictionary<string, object> { ["type"] = "string" }
                    },
                    ["required"] = new[] { "city" },
                },
                OutputSchema: new Dictionary<string, object>
                {
                    ["type"] = "object",
                    ["properties"] = new Dictionary<string, object>
                    {
                        ["tempC"] = new Dictionary<string, object> { ["type"] = "number" },
                        ["condition"] = new Dictionary<string, object> { ["type"] = "string" },
                    },
                },
                Operator: identity.Did,
                TestVectors:
                [
                    new TestVector(
                        Input: new Dictionary<string, object> { ["city"] = "Berlin" },
                        ExpectedOutput: new Dictionary<string, object>
                        {
                            ["tempC"] = 18,
                            ["condition"] = "cloudy",
                        },
                        Description: "Berlin weather lookup"
                    ),
                ]
            ),
        ],
    }
);

// Invoke the tool
var result = await ctx.InvokeToolAsync("weather", new Dictionary<string, object> { ["city"] = "Berlin" });
Console.WriteLine($"Weather result: {result}");

File: bindings/go/examples/tool_invocation.go
// Tool invocation: register a tool with test vectors and invoke it.
package main

import (
	"fmt"
	"log"

	scp "github.com/limn/scp-go"
)

func main() {
	if err := scp.Init(); err != nil {
		log.Fatal(err)
	}
	defer scp.Shutdown()

	identity, err := scp.NewIdentity("platform")
	if err != nil {
		log.Fatal(err)
	}

	ctx, err := scp.NewContext(identity, scp.ContextParams{
		Ceiling: []string{"msg:send", "msg:receive", "tool:invoke"},
		Tools: []scp.ToolDefinition{
			{
				Name:        "weather",
				Description: "Get current weather for a city",
				InputSchema: map[string]any{
					"type":       "object",
					"properties": map[string]any{"city": map[string]any{"type": "string"}},
					"required":   []string{"city"},
				},
				OutputSchema: map[string]any{
					"type": "object",
					"properties": map[string]any{
						"tempC":     map[string]any{"type": "number"},
						"condition": map[string]any{"type": "string"},
					},
				},
				Operator: identity.DID(),
				TestVectors: []scp.TestVector{
					{
						Input:          map[string]any{"city": "Berlin"},
						ExpectedOutput: map[string]any{"tempC": 18, "condition": "cloudy"},
						Description:    "Berlin weather lookup",
					},
				},
			},
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	defer ctx.Close()

	// Invoke the tool
	result, err := ctx.InvokeTool("weather", map[string]any{"city": "Berlin"})
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Weather result:", result)
}
